Add swift monad post

endoze opened the pull request:
This commit adds the first draft of the 'Swift Optionals Are Monads
Too!' post.

endoze committed:
Add swift monad post

This commit adds the first draft of the 'Swift Optionals Are Monads
Too!' post.

calebkm commented:
w00t w000t! This is great, nice job taking the initiative and pumping this out. I'd say it's still _slightly_ over my head, but I'm not very smart so that's not too surprising.

calebkm commented on posts/Swift-Optionals-Are-Monads-Too.md:
````diff
@@ -0,0 +1,264 @@
+}
+```
+
+Reading the first comment in the source code gives us an awesome definition of what an Optional really is. 
````
My only comment is that the word `awesome` here is maybe just slightly hyperbolic. Maybe `awesome` is a "cooler" word, but I think the sentence flows a little nicer with a `great` or a `perfect`.

northband replied:
Or the word `clear` - "...in the source code gives us a clear definition"

endoze replied:
I'm just that excited about the definition. I could run with `great`, `perfect`, or `clear`.

northband commented on posts/Swift-Optionals-Are-Monads-Too.md:
```diff
@@ -0,0 +1,264 @@
+
+> A type that can represent either a Wrapped value or nil, the absence of a value.
+
+Put in other words, an Optional is a container that wraps the concept of having a value or not.
```
`Put` may be optional - it flows well as 'In other words,'.

northband commented on posts/Swift-Optionals-Are-Monads-Too.md:
```diff
@@ -0,0 +1,264 @@
+
+Put in other words, an Optional is a container that wraps the concept of having a value or not.
+
+So why is this useful? Because Swift is a strongly typed language it makes us code for both cases of an Optional, having a value (.Some) or no value at all (.None). When we declare variables with let or var and make them Optional, we are telling the compiler that our variable may contain a value or not. And any subsequent access to our variable will require us to check that it contains a value be ...
```
This may read better:

So why is this useful? Because Swift is a strongly typed language `that` makes us code for both cases of an Optional.  That is to say, having a value (.Some) or no value at all (.None). When we declare variables with let or var and make them Optional, we are telling the compiler that our variable may contain a value or not. `Any` subsequent access to our variable will require us to check that it contains a value before performing actions on the wrapped value.

northband commented on posts/Swift-Optionals-Are-Monads-Too.md:
````diff
@@ -0,0 +1,264 @@
+```
+---
+
+# What the Functor
````
May want to add a question mark: What the Functor?

northband commented on posts/Swift-Optionals-Are-Monads-Too.md:
```diff
@@ -0,0 +1,264 @@
+
+# What the Functor
+
+So what is a functor? It sounds complicated and only adding to it's complicated nature, it's roots are in Mathematics (Category Theory to be a bit more specific). But let's try and define it in terms a programmer understands instead of mathematical terms.
```
Maybe 'adding to it's `complex` nature,...'

northband commented on posts/Swift-Optionals-Are-Monads-Too.md:
````diff
@@ -0,0 +1,264 @@
+}
+```
+
+If we break down Optional's implementation of map, it's a function that accepts another function as it's only argument. This other function, we'll call it f, accepts an object of the type T and returns an object of the type U. And finally, map returns U? which is an Optional of the type U.
````
👍

northband commented on posts/Swift-Optionals-Are-Monads-Too.md:
```diff
@@ -0,0 +1,264 @@
+
+> An applicative functor is any type that defines apply.
+
+Looking back at the declaration for Optional, we don't find any methods defining apply. Well hmm, by default Optionals aren't applicative functors. But wait! We can add an extension to this built in type and make it work. 
```
It's totally acceptable but you don't need to use words like 'Well hmm,' you can just do 'Hmm'.  Or 'But wait!', could be 'Wait!'.  Again this is just a style thing.

northband commented on posts/Swift-Optionals-Are-Monads-Too.md:
````diff
@@ -0,0 +1,264 @@
+}
+```
+
+Looking at the implementation of apply kind of melts your brain, so let's break it down and explain it in simpler terms. First we should notice a very important detail buried in the middle of the function. f's type is (Wrapped -> U)?. f is an Optional! This little interesting tidbit is the key difference between map and apply. So looking into the implementation we see that we are calling flatMap  ...
````
👍

northband commented on posts/Swift-Optionals-Are-Monads-Too.md:
```diff
@@ -0,0 +1,264 @@
+
+This allows us to choose two completely different strings based on whether our `numberOfFriends` variable has a value or not.
+
+One last example of a great usage for `map`. What if we want to read the contents of a url?
```
Maybe:

Another example of a great usage...

This may cut down on simple minds thinking this is not the actual last case where it's useful ;-)

northband commented on posts/Swift-Optionals-Are-Monads-Too.md:
````diff
@@ -0,0 +1,264 @@
+webContents.map { print($0) }
+```
+
+If I was a bit more mischievous, I could leave the example like that and go on with the post. But the most astute reader who really gets these concepts would point out the mistake in the code sample above. In case you didn't catch it, that's ok. The problem with our code above is that map accepts a closure/function that does not return an `Optional`. `NSData`'s initializer that accepts an `NSURL` ...
````
👍

northband commented on posts/Swift-Optionals-Are-Monads-Too.md:
````diff
@@ -0,0 +1,264 @@
+webContents.map { print($0) }
+```
+
+### How do we really apply
````
Maybe:

Real world uses of apply

endoze replied:
I was aiming for a little bit on the light side instead of sticking with something specific and to the point.

northband commented on posts/Swift-Optionals-Are-Monads-Too.md:
```diff
@@ -0,0 +1,264 @@
+
+# Wrapping Up
+
+Hopefully by now you can see how useful functors, applicative functors, and monads can really be. Apple making Optionals work as functors and monads really make writing intentful code a cinch. And just to keep your brain thinking about this crazy topic, Swift's Array type is also a functor and a monad!
```
Maybe:

Apple's decision to make Optionals work as functors and monads really allows writing confident code a cinch. Lastly, just to keep your brain thinking about this crazy topic, Swift's Array type is also a functor and a monad!

endoze replied:
I'm no fan of using `lastly`, otherwise I like the suggested edits.

northband commented:
This is really good @endoze - nice work!  I just made red marks to shorten things up a bit.  Take them as you will.

endoze committed:
Update post with suggested edits

In order to make this post read well, this commit updates it with
suggested edits.

endoze committed:
Update code samples and comments

In order to ensure the code samples run in a playground, this commit
updates variable names and comments.

endoze committed:
Add link to post in README

In order to make viewing the posts easier, this commit updates the
readme with links to the posts.
